src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size;
+
+    private void ensureCapacity() {
+        if (size == elements.length) {
+            int newCapacity = (int)(elements.length * 1.5);
+            elements = Arrays.copyOf(elements, newCapacity);
+        }
+    }
+
+    private void ensureCapacity(int DEFAULT_CAPACITY) {
+        if (DEFAULT_CAPACITY > elements.length) {
+            int newCapacity = elements.length + elements.length / 2;
+            if (newCapacity < DEFAULT_CAPACITY) {
+                newCapacity = DEFAULT_CAPACITY;
+            }
+            elements = Arrays.copyOf(elements, newCapacity);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        ensureCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(size + list.size());
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        T oldValue = (T) elements[index];
+        int numMoved = size - index - 1;
+        if (numMoved > 0) {
+            System.arraycopy(elements, index + 1, elements, index, numMoved);
+        }
+        elements[--size] = null;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        // Find the index of the element to remove
+        int index = (int) element;
+        if (index == -1) {
+            throw new NoSuchElementInArrayListException("Element not found: " + element);
+        }
+
+        return remove(index);
```

Comment:
The remove(T element) method is incorrect. Casting 'element' to an int (line 85) is not the proper way to find the index of an element. Instead, you should iterate over the array to find the actual index where the element is located.